subprojects/base-services/src/main/java/org/gradle/internal/classloader/VisitableURLClassLoader.java

```diff
@@ -33,6 +36,28 @@
         }
     }
 
+    private final EnumMap<UserData, Object> userData = new EnumMap<UserData, Object>(UserData.class);
```

Comment:
I think an enummap + synchronized here is not any better than using a concurrent hash map. I'd go with the latter, so we don't need to keep an enum of keys here, giving us better separation of concerns.

Reply:
The `EnumMap` has the main advantage of having descriptive keys and optimized lookup/size. Also a `ConcurrentMap` will not prevent the same value from being computed concurrently, which we want to avoid (or we could have race conditions and linkage errors).